graylog2-web-interface/src/views/components/visualizations/ChartData.js

```diff
@@ -10,64 +10,72 @@ export type ChartDefinition = {
   name: string,
   x?: Array<string>,
   y?: Array<any>,
+  z?: Array<Array<any>>,
 };
 
-export type ChartData = [any, Array<Key>, Array<any>]
+export type ChartData = [any, Array<Key>, Array<any>, Array<Array<any>>]
 export type ExtractedSeries = Array<ChartData>;
 
 export type KeyJoiner = (Array<any>) => string;
 
-export type Generator = (string, string, Array<string>, Array<any>, number, number) => ChartDefinition;
+export type Generator = (string, string, Array<string>, Array<any>, Array<Array<any>>, number, number) => ChartDefinition;
 
 const _defaultKeyJoiner = keys => keys.join('-');
 
-export const extractSeries = (keyJoiner: KeyJoiner = _defaultKeyJoiner) => {
-  return (results: Rows): ExtractedSeries => {
-    // $FlowFixMe: Somehow flow is unable to infer that the result consists only of Leafs.
-    const leafs: Array<Leaf> = results.filter(row => (row.source === 'leaf'));
+const _defaultChartGenerator = (type, name, labels, values): ChartDefinition => ({ type, name, x: labels, y: values });
 
-    const x = leafs.map(({ key }) => key);
+const _flattenLeafs = (leafs: Array<Leaf>, matcher: string => boolean = source => source.endsWith('leaf')) => {
+  return flatten(leafs.map(l => l.values.filter(value => matcher(value.source)).map(v => [l.key, v])));
```

Comment:
Wouldn't it be better if we would pass in `value` instead of `value.source` to the `matcher` function? It would offer better flexibility (maybe we want to filter based on key or values?) and does not make calling significantly longer (the matcher would be specified as `({ source }) => ...` instead of `source => ...`.

Reply:
Now I get your point regarding the flexibility. I did not saw the benefits of using the full value in the first place.